OpenRA.Mods.Common/UpdateRules/Rules/20201213/RemoveResourceType.cs

```diff
@@ -0,0 +1,131 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace OpenRA.Mods.Common.UpdateRules.Rules
+{
+	public class RemoveResourceType : UpdateRule
+	{
+		public override string Name { get { return "Remove ResourceType definitions."; } }
+
+		public override string Description
+		{
+			get
+			{
+				return "The ResourceType trait has been removed, and resource definitions moved to the\n" +
+					"ResourceLayer, EditorResourceLayer, ResourceRenderer, and PlayerResources traits.";
+			}
+		}
+
+		MiniYaml resourceLayer;
+		MiniYaml resourceRenderer;
+		MiniYaml values;
+
+		public override IEnumerable<string> BeforeUpdate(ModData modData)
+		{
+			resourceLayer = new MiniYaml("");
+			resourceRenderer = new MiniYaml("");
+			values = new MiniYaml("");
+			yield break;
+		}
+
+		public override IEnumerable<string> AfterUpdate(ModData modData)
+		{
+			if (resourceLayer.Nodes.Any())
+				yield return "Add the following definitions to your ResourceLayer and EditorResourceLayer definitions:\n\t" +
+					"RecalculateResourceDensity: true\n\t" +
+					resourceLayer.ToLines("ResourceTypes").JoinWith("\n\t");
+
+			if (resourceLayer.Nodes.Any())
+				yield return "Add the following definitions to your ResourceRenderer definition:\n\t" +
+					resourceRenderer.ToLines("ResourceTypes").JoinWith("\n\t");
+
+			if (values.Nodes.Any())
+				yield return "Add the following definition to your ^BasePlayer definition:\n\t" +
+					"PlayerResources:\n\t\t" +
+					values.ToLines("ResourceValues").JoinWith("\n\t\t");
+
+			if (resourceLayer.Nodes.Any())
+				yield return "Support for AllowUnderActors, AllowUnderBuildings, and AllowOnRamps have been removed.\n" +
+					"You must define a custom ResourceLayer subclass if you want to customize the default behaviour.";
```

Comment:
Yes. These options were counterproductive because the actual rules that we need to implement are too subtle to encode in a single flag (e.g. TS allows resources only on ramp types 0-4, but never 5-20, and veins requires several sentences to explain). I felt it would be better to hardcode a sensible neutral default and require all mods to subclass the layer than to implement flags for `IsTSTiberium` and `IsTSVeins`.